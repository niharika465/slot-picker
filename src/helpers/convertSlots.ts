import { SlotTime } from 'types';

type Slot = {
  displayDate: string;
  displayTime: string;
  displayTimeEnd: string;
  startTimeUtc: number;
  endTimeUtc: number;
};

type GroupedSlots = {
  [key: string]: SlotTime;
};

const convertSlots = (slots: Slot[]): GroupedSlots => {
  const groupedSlots: GroupedSlots = {};

  slots.forEach((slot) => {
    const { displayDate, displayTime } = slot;

    if (!groupedSlots[displayDate]) {
      groupedSlots[displayDate] = [];
    }

    groupedSlots[displayDate].push({ displayTime });
  });

  return groupedSlots;
};

export default convertSlots;
