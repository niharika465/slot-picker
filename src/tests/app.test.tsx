import { render, screen, fireEvent } from '@testing-library/react';
import App from '../app';
import convertSlots from '../helpers/convertSlots';
import getDaysAndWeekdays from '../helpers/getWeekday';

// Mocking dependencies
jest.mock('../helpers/convertSlots', () => jest.fn());
jest.mock('../helpers/getWeekday', () => jest.fn());

// Mock class for IntersectionObserver
class IntersectionObserverMock {
  private callback: IntersectionObserverCallback;
  private observedElements: Element[] = [];

  constructor(callback: IntersectionObserverCallback) {
    this.callback = callback;
  }

  observe(element: Element) {
    this.observedElements.push(element);
    // Trigger the callback
    this.callback(
      [
        {
          isIntersecting: true,
          boundingClientRect: element.getBoundingClientRect(),
          intersectionRatio: 1,
          target: element,
          time: 0,
        } as IntersectionObserverEntry,
      ],
      this // Pass the mock instance as the second argument
    );
  }

  unobserve(element: Element) {
    this.observedElements = this.observedElements.filter((e) => e !== element);
  }

  disconnect() {
    this.observedElements = [];
  }

  // Added these properties incase typescript doesn't recognize them
  get root() {
    return null;
  }

  get rootMargin() {
    return '';
  }

  get thresholds() {
    return [];
  }

  takeRecords() {
    return [];
  }
}

// global scope for the observer
// eslint-disable-next-line @typescript-eslint/no-explicit-any
(global as any).IntersectionObserver = IntersectionObserverMock;

const mockSlots = {
  '2023-10-01': [{ displayTime: '09:00' }, { displayTime: '10:00' }],
  '2023-10-02': [{ displayTime: '09:00' }, { displayTime: '10:00' }],
};

beforeEach(() => {
  (convertSlots as jest.Mock).mockReturnValue(mockSlots);
  (getDaysAndWeekdays as jest.Mock).mockReturnValue([
    { day: '01', weekday: 'Mon', date: '2023-10-01' },
    { day: '02', weekday: 'Tue', date: '2023-10-02' },
  ]);
});

afterEach(() => {
  jest.clearAllMocks();
});

test('renders the component successfully', () => {
  const { asFragment } = render(<App />);
  // match withe snapshot
  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByText('Pick a date')).toBeInTheDocument();
});

test('the Book button should be disabled when no slot timing is selected', () => {
  render(<App />);
  const button = screen.getByText('Book');
  expect(button).toHaveClass('button--disabled');
});

test('the Book button should be enabled when a slot timing is selected', () => {
  render(<App />);

  // selecting a slot: triggered onclick
  fireEvent.click(screen.getByText('01'));

  // selecting a time: triggered onclick
  fireEvent.click(screen.getByText('09:00'));

  const button = screen.getByText('Book');
  expect(button).not.toHaveClass('button--disabled');
});

test('when Book button is clicked: goes back to initial state', () => {
  render(<App />);

  // selecting a slot: triggered onclick
  fireEvent.click(screen.getByText('01'));

  // selecting a time: triggered onclick
  fireEvent.click(screen.getByText('09:00'));

  const bookButton = screen.getByText('Book');
  // Click the Book button
  fireEvent.click(bookButton);
  // after clicking the button , everything goes back to initial state and button is disabled again
  expect(bookButton).toHaveClass('button--disabled');
});

test.skip('handleReset should be triggered on Cancel button click', () => {
  // selecting a slot: triggered onclick
  fireEvent.click(screen.getByText('01'));

  // selecting a time: triggered onclick
  fireEvent.click(screen.getByText('09:00'));

  const cancelButton = screen.getByText('Cancel');
  fireEvent.click(cancelButton);

  // After cancel, the selected slot and time should be reset and the button is disabled again
  expect(cancelButton).toHaveClass('button--disabled');
});
