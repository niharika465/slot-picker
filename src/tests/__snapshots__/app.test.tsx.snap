// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component successfully 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div>
      <div
        class="heading"
      >
        Pick a date
      </div>
      <div
        class="slider"
      >
        <div
          class="navigation-btn navigation-btn--disabled"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-tcjmn5-MuiSvgIcon-root"
            data-testid="ArrowBackIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20z"
            />
          </svg>
        </div>
        <div
          class="slide-container"
        >
          <div
            class="slide-item"
          >
            <div
              class=""
            >
              01
            </div>
            <div
              class="slide-item__day"
            >
              Mon
            </div>
          </div>
          <div
            class="slide-item"
          >
            <div
              class=""
            >
              02
            </div>
            <div
              class="slide-item__day"
            >
              Tue
            </div>
          </div>
        </div>
        <div
          class="navigation-btn navigation-btn--disabled"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-tcjmn5-MuiSvgIcon-root"
            data-testid="ArrowForwardIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="button__controls"
    >
      <div
        class="button button--disabled"
      >
        Book
      </div>
      <div
        class="button button--disabled"
      >
        Cancel
      </div>
    </div>
  </div>
</DocumentFragment>
`;
